Podcast/Player.swift

```diff
@@ -9,59 +9,189 @@
+    /// - Parameters:
+    ///   - url: The url to fetch and play from.
+    func prepareToPlay(url: URL) {
+        if currentURL != url || playerStatus != .preparingToPlay {
+            // create a new player if this is a new URL or we aren't currently preparing a URL
+            playerStatus = .preparingToPlay
+            currentAVPlayer = AVPlayer(url: url)
+            currentAVPlayer?.addObserver(self, forKeyPath: #keyPath(AVPlayer.status), options: [.old, .new], context: &playerContext)
+        }
+    }
+    
+    /// Plays the current track and sets the Player status to .playing if successful
     func play() {
-        prepareToPlay()
-        player?.play()
+        if let player = currentAVPlayer {
+            player.play()
+            playerStatus = .playing
+            player.addPeriodicTimeObserver(forInterval: CMTimeMakeWithSeconds(1.0, Int32(NSEC_PER_SEC)), queue: DispatchQueue.main, using: { [weak self] time in
+                self?.delegate?.playerDidUpdateTime()
+            })
+        }
     }
     
+    /// Pauses playback of the current track and sets the Player status to .paused if successful
     func pause() {
-        player?.pause()
+        if let player = currentAVPlayer {
+            player.pause()
+            playerStatus = .paused
+        }
     }
     
+
+    /// Toggles playback of the Player. If the player is in the preparingToPlay state and cannot yet play,
+    /// this toggles whether the player will autoplay once it is ready.
     func togglePlaying() {
-        isPlaying ? pause() : play()
+        switch playerStatus {
+        case .empty:
+            break
+        case .preparingToPlay:
+            shouldAutoPlay = !shouldAutoPlay
+        case .playing:
+            pause()
+        case .paused:
+            play()
+        case .failed:
+            break
+        }
     }
     
-    // MARK: - AVAudioPlayerDelegate
+    /// Skips a variable amount of seconds in the current audio track. Skips forward or backward based on a positive or negative seconds value.
+    /// - Parameters:
+    ///   - seconds: the amount of time in seconds to skip. If negative, skips backward.
+    func skip(seconds: Float64) {
+        if let player = currentAVPlayer {
+            let newTime = CMTimeAdd(player.currentTime(), CMTimeMakeWithSeconds(seconds, Int32(NSEC_PER_SEC)))
+            player.seek(to: newTime)
+            self.delegate?.playerDidUpdateTime()
+        }
+    }
     
     
-    func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully flag: Bool) {
-        pause()
+    /// Returns the progress for the Player's current track.
+    /// - Returns: a Float in the range [0.0,1.0] specifying what the current location of playback is relative to the entire track.
+    func getProgress() -> Float {
+        if playerStatus == .empty || playerStatus == .failed || currentAVPlayer == nil || currentAVPlayer?.currentItem == nil {
```

Comment:
Redundant case `currentAVPlayer == nil` since the optional chaining on the next condition will yield nil if the player is nil.